src/ExactDiagonalization/basis_fock.jl

```diff
@@ -0,0 +1,181 @@
+"""
+    build_basis(addr::AbstractFockAddress)
+    build_basis(::Type{<:AbstractFockAddress}) -> basis
+
+Return all possible Fock states of a given type as a vector. This method is _much_ faster
+than `build_basis(::AbstractHamiltonian, ...)`, but does not take matrix blocking into
+account. This version of `build_basis` accepts no additional arguments.
+
+All address types except [`OccupationNumberFS`](@ref Main.Rimu.OccupationNumberFS) are
+supported.
+
+Returns a sorted vector of length equal to the [`dimension`](@ref) of `addr`.
+"""
+function build_basis(addr::AbstractFockAddress)
+    return build_basis(typeof(addr))
+end
+
+###
+### BoseFS
+###
+# Algorithm:
+# Given an address type of N particles in M modes,
+# For all n_1 ∈ (0 ... N)
+# - put n_1 particles in the first mode
+# - recursively call the function to put the remaining (N-n_1) particles into (M-1) modes.
+# The recursion is parallelised through spawning tasks.
+
+# this is equivalent to `dimension(BoseFS{n,m})`, but does not return a `BigInt`.
+_bose_dimension(n, m) = binomial(n + m - 1, n)
+
+function build_basis(::Type{<:BoseFS{N,M}}) where {N,M}
+    T = typeof(near_uniform(BoseFS{N,M}))
+    result = Vector{T}(undef, _bose_dimension(N, M))
+    _bose_basis!(result, (), 1, N, Val(M), 2 * Threads.nthreads())
+    return result
+end
+
+# Multithreaded version - attempts to spawn `n_tasks` tasks.
+@inline function _bose_basis!(
+    result::Vector, postfix, index, remaining_n, ::Val{M}, n_tasks::Int
+) where {M}
+    @sync if M < 5 || remaining_n ≤ 1 || n_tasks ≤ 0
```

Comment:
I think I understand how this algorithm works, but it is not clear to me why the work is split up the way it is. Is it just for performance or is there some other reason?
E.g.: Why is `M < 5` the condition here?

Reply:
Spawning threads on small addresses is not worth the overhead. I did some benchmarking and `M < 5` seemed like an ok place to put the threshold.